### PR TITLE
fix: placeholders now check for empty content compared to their "true" empty state

### PR DESCRIPTION
**Pull request by nperez0111**
## Changes Overview
This resolves a bug where the placeholder extension was not counting a document that contains more complex default content

## Implementation Approach

I did some testing with the placeholder extension, couldn't be bothered to write whole new tests for it:
```diff
diff --git a/demos/src/Extensions/Placeholder/React/index.jsx b/demos/src/Extensions/Placeholder/React/index.jsx
index 970353980..9d5ae3389 100644
--- a/demos/src/Extensions/Placeholder/React/index.jsx
+++ b/demos/src/Extensions/Placeholder/React/index.jsx
@@ -1,5 +1,7 @@
 import './styles.scss'
 
+import { Node } from '@tiptap/core'
+import Document from '@tiptap/extension-document'
 import Placeholder from '@tiptap/extension-placeholder'
 import { EditorContent, useEditor } from '@tiptap/react'
 import StarterKit from '@tiptap/starter-kit'
@@ -8,7 +10,19 @@ import React from 'react'
 export default () => {
   const editor = useEditor({
     extensions: [
-      StarterKit,
+      Node.create({
+        name: 'page',
+        content: 'block+',
+        renderHTML({ HTMLAttributes }) {
+          return ['p', HTMLAttributes, 0]
+        },
+      }),
+      Document.extend({
+        content: 'page+',
+      }),
+      StarterKit.configure({
+        document: false,
+      }),
       Placeholder.configure({
         // Use a placeholder:
         placeholder: 'Write something …',

```

## Testing Done

^ above
## Checklist
- [ ] I have created a [changeset](https://github.com/changesets/changesets) for this PR if necessary.
- [ ] My changes do not break the library.
- [ ] I have added tests where applicable.
- [ ] I have followed the project guidelines.
- [ ] I have fixed any lint issues.

### CODE REVIEW

**Review by bdbch — APPROVED** [review]
LGTM! I think we just need a changeset file for this I guess

File: packages/extension-placeholder/src/placeholder.ts [hunk 1]
```diff
@@ -31,15 +31,6 @@ export interface PlaceholderOptions {
...
     | string
 
-  /**
-   * **Used for empty check on the document.**
-   *
-   * If true, any node that is not a leaf or atom will be considered for empty check.
-   * If false, only default nodes (paragraphs) will be considered for empty check.
-   * @default false
-   */
-  considerAnyAsEmpty: boolean
```
[the remark this answers is not in the document]
Comment (bdbch) [reply packages/extension-placeholder/src/placeholder.ts:41]: Good point - didn't thought about this. @nperez0111 what do you think? should we push this to a major and make sure to note that the option is gone?

Or should we keep it somehow if possible?

[the remark this answers is not in the document]
Comment (nperez0111) [reply packages/extension-placeholder/src/placeholder.ts:41]: Ah sorry I didn't fully document why I think that this was actually OK to remove.

So the reason that was needed was actually masking a fact of the implementation (which was checking if the first child of the doc was the default) with this:
```js
            const isValidNode = this.options.considerAnyAsEmpty
              ? true
              : firstChild && firstChild.type.name === doc.type.contentMatch.defaultType?.name
```

Now, this check is removed because it has become unnecessary in my implementation.

What enabled that?
Previously there was a change made for performance to do a more performant check to see if the editor was empty or not (as opposed to the previous editor.isEmpty): https://github.com/ueberdosis/tiptap/commit/6e5d19f65d98669741973a2812ae13afeea65672

Then we added a more complex check to try to enable placeholders on any node type: https://github.com/ueberdosis/tiptap/pull/4335

But this more complex check did not account for nodes that contain other empty nodes inside of them (issue being `firstChild.nodeSize <= 2` which fails for a node that contains an empty node)

I actually had recently changed the `editor.isEmpty` implementation to be much more performant: https://github.com/ueberdosis/tiptap/commit/f4a7dbaf5c8ab8e6f3bcd17db2cb0802d1f95e22

So, now we could skip all of the complexity around this and go back to the original implementation, but with the nice performance benefit of only check if the editor is empty once.


## TL;DR

That option is no longer necessary anymore. And it is not a breaking change because even if a user specified that value or not, it would not break them and show/hide a placeholder.

[the remark this answers is not in the document]
Comment (nperez0111) [reply packages/extension-placeholder/src/placeholder.ts:41]: Unsure about updating docs though. Technically the option is still there in the current release...

One downside of keeping the docs in a separate repo...

[the remark this answers is not in the document]
Comment (svenadlung) [reply packages/extension-placeholder/src/placeholder.ts:41]: Oh, thank you for your explanation. That makes so much sense. I didn't realize that `isValidNode` and `considerAnyAsEmpty` are only important for checking the empty state of the document, rather than also being considered for empty nodes in general. But I ran a few tests and everything seems to be fine + re-reading the specific code and your explanation made it clear that it was only for checking the empty document.

[the remark this answers is not in the document]
Comment (svenadlung) [reply packages/extension-placeholder/src/placeholder.ts:41]: Yeah re docs: Maybe prepare a PR there and we need to keep track of when it gets released?

[the remark this answers is not in the document]
Comment (rChaoz) [reply packages/extension-placeholder/src/placeholder.ts:41]: Hi, @nperez0111! Just came across this PR after some digging that ended up with cloning tiptap and doing `git log v2.3.1..v2.5.7 -S considerAnyAsEmpty` :p

I just have a quick note and question. The note is that, while this is not a breaking change for JS users, it is for TS users, as I just updated tiptap and suddently I came across this error, which I did not understand as I did a minor version bump:

```
Error: Object literal may only specify known properties, and 'considerAnyAsEmpty' does not exist in type 'Partial<PlaceholderOptions>'. (ts)
```

Naturally I scanned the changelog and didn't find anything. The code would still run, but the CI fails, so I'd say it is a bit breaking, not a big problem but confusing nonetheless.

And now for the question: I don't quite remember why I set the thing to be true in the first place, but I'm certain that it didn't work otherwise. So, is the new implementation behaving like the old implementation with `considerAnyAsEmpty = false` or `true`? From my understanding it's the latter, just want to make sure 🙂

[the remark this answers is not in the document]
Comment (nperez0111) [reply packages/extension-placeholder/src/placeholder.ts:41]: The considerAnyAsEmpty was a kludge for the isNodeEmpty function not being able to accurately report if content was empty. So not the behavior is as though considerAnyAsEmpty = true but actually catches more cases than just that. But you should try this yourself. 

I could've easily left the type there and mark it deprecated and just totally ignore the value of it. I'll add it back so it doesn't "break" others but it should be removed.

[the remark this answers is not in the document]
Comment (rChaoz) [reply packages/extension-placeholder/src/placeholder.ts:41]: Makes sense, thanks for the confirmation

**Comment by nyacg** [comment]
FYI this has caused a regression for us. Now the placeholder shows for the parent component and the child if the child is empty. We have `showOnlyCurrent` set to `false` and then a check for `isAchor` 
image

**Comment by nyacg** [comment]
Added a  check for `childCount` in my `placeholder` function and this seems to match previous behaviour 
```javascript
if (!hasAnchor || node.childCount) {
    return "";
}
````

**Comment by nperez0111** [comment]
@nyacg I'm looking at this & I can't seem to reproduce the issue you are describing, can you share your placeholder configuration before & after?

Your screenshot is what I would expect if `showOnlyCurrent` was `false` so I'm unsure if that is something I should resolve? But it was not totally clear if this was before or after your workaround

**Comment by nperez0111** [comment]
I've opened a PR that would contain this fix once we can get to the bottom of it: https://github.com/ueberdosis/tiptap/pull/5409

**Comment by nyacg** [comment]
Sure, here's the full diff

image

I just restored the behaviour from this change https://github.com/ueberdosis/tiptap/pull/5278/files#diff-4fe0b8cec20e4964b0f43fc19db9c132f260ca2f89a625b48a9fcbf748b7eed4L119

It seems like it doesn't count ReactNodeView nodes as content by default 


Not an important breaking change and we're happy to work around it, just thought I'd document for anyone that also notices this

File: packages/extension-placeholder/src/placeholder.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nperez0111) [reply packages/extension-placeholder/src/placeholder.ts:41]: I added the deprecated type back with v2.5.9

**Comment by hivokas** [comment]
I was experiencing a very similar issue to what @nyacg mentioned above, but in my case with VueNodeView. Upgrading to the latest version fixed the problem for me (thanks to https://github.com/ueberdosis/tiptap/pull/5409 I believe).
